### PR TITLE
Use a connection pool for gRPC client calls from jobservice to armada-server

### PR DESCRIPTION
**Pull request by ClifHouck**
Addresses [BATCH-225](https://gr-oss.atlassian.net/browse/BATCH-225) 

Uses my [custom forked version](https://github.com/ClifHouck/grpc-go-pool) of [grpc-go-pool](https://github.com/processout/grpc-go-pool) to add a pooled grpc armada-server client to the jobservice.

[BATCH-223]: https://gr-oss.atlassian.net/browse/BATCH-223?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

┆Issue is synchronized with this [Jira Task](https://gr-oss.atlassian.net/browse/BATCH-225) by [Unito](https://www.unito.io)

[BATCH-225]: https://gr-oss.atlassian.net/browse/BATCH-225?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: go.sum
Comment (suprjinx) [inline go.sum:66]: should this be referencing your github?

**Review by suprjinx — APPROVED** [review]

File: go.sum
Comment (dave-gantenbein) [reply go.sum:66]: +1

File: internal/jobservice/application.go
Comment (dave-gantenbein) [inline internal/jobservice/application.go:38]: #testme

File: internal/jobservice/events/client.go
Comment (dave-gantenbein) [inline internal/jobservice/events/client.go:1]: Should we be testing the new methods here?

File: go.sum
Comment (ClifHouck) [reply go.sum:66]: Depends. The root of my fork does not seem to be actively maintained. I've made modifications to the pool that make more sense for grpc connections. For instance, grpc connections are safe to use across many gorountines at once, so a client grabbing one shouldn't remove it from the pool. The pool just round-robins access to clients to load balance their use.

File: internal/jobservice/events/client.go
Comment (ClifHouck) [reply internal/jobservice/events/client.go:1]: Yes, will do.

File: go.sum
Comment (kannon92) [reply go.sum:66]: As a OSS team I think the right approach would be to open this issue on their repo and maybe propose a fix to their code.  

I don't think forking and including your fork is a pattern we want to encourage from the open source team.

Comment (ClifHouck) [reply go.sum:66]: Generally I agree with you, but the context of this repo matters. 

I'm happy to move this fork into `armadaproject` or just move the two source files into `armadaproject/armada`. To reiterate: the root of my fork is not actively maintained, as in it hasn't had a commit for *years*. It's likely maintained by a single person who has effectively abandoned the project. The changes I've made here also totally change the semantics and functionality of the pool, so unlikely to be merged upstream.

File: internal/jobservice/events/client.go
Comment (ClifHouck) [reply internal/jobservice/events/client.go:1]: Addressed in 9d0633f

File: internal/jobservice/application.go
Comment (ClifHouck) [reply internal/jobservice/application.go]: Addressed in [a465a04](https://github.com/armadaproject/armada/pull/2420/commits/a465a04f4cf23b8f8cebdaabc1c24c657d0a8afd)

File: internal/common/grpc/grpcpool/LICENSE
Comment (kannon92) [inline internal/common/grpc/grpcpool/LICENSE:1]: I really don't know what to do about this.  

I believe Armada-Operator is Licensed as Apache.  How does mixing different Software licenses work?

File: internal/jobservice/application.go
Comment (kannon92) [inline internal/jobservice/application.go:46]:
```suggestion
func RectifyConfig(config *configuration.JobServiceConfiguration) {
```

I don't think you ever return an error here.

File: internal/jobservice/application.go
Comment (kannon92) [inline internal/jobservice/application.go:73]: I think this panic would never be hit.

**Review by kannon92 — COMMENTED** [review]
I think we should answer the mixing of licenses questions before we merge this.  

I'm not a lawyer but that screams an issue to me.  

The code LGTM though.  Just one nit about the fact that you don't return errors from the config function so we probably don't need the return type.

File: internal/common/grpc/grpcpool/LICENSE
Comment (dave-gantenbein) [reply internal/common/grpc/grpcpool/LICENSE:1]: https://law.stackexchange.com/questions/6081/can-i-bundle-mit-licensed-components-in-a-apache-2-0-licensed-project#:~:text=Both%20the%20Apache%20License%20and,as%20required%20by%20the%20license.

Comment (kannon92) [reply internal/common/grpc/grpcpool/LICENSE:1]: well TIL.  

I guess its fine then?

Comment (dave-gantenbein) [reply internal/common/grpc/grpcpool/LICENSE:1]: Ship it.

File: internal/common/grpc/grpcpool/pool_test.go
Comment (dave-gantenbein) [inline internal/common/grpc/grpcpool/pool_test.go:228]: If its not relevant, lets remove it.

File: internal/jobservice/application.go
Comment (ClifHouck) [reply internal/jobservice/application.go:46]: I don't, but in the future I imagine there'd be a need to. This is only checking config variables I added in this PR but eventually we should probably be doing some kind of validation of all configuration options.

File: internal/jobservice/application.go
Comment (ClifHouck) [reply internal/jobservice/application.go:73]: It won't but returning an error from `RectifyConfig` in the future is intended to indicate an unrecoverable error with `config`. Just putting in the correct behavior now.

File: internal/common/grpc/grpcpool/pool_test.go
Comment (ClifHouck) [reply internal/common/grpc/grpcpool/pool_test.go]: Addressed in 440ce3e

File: internal/jobservice/application.go
Comment (kannon92) [reply internal/jobservice/application.go:46]: This seems to me to be YAGNI.  If we expand to do validation we could add returning errors but as of right now we don't need the return.  I'll resolve though if you think that best.

**Review by kannon92 — APPROVED** [review]

**Review by suprjinx — APPROVED** [review]

File: internal/common/grpc/grpcpool/pool.go
Comment (richscott) [inline internal/common/grpc/grpcpool/pool.go:35]: Can this commented-out line be removed?

**Review by richscott — APPROVED** [review]
I noticed a line of old commented-out code, but otherwise don't want to hold this up - looks good.

File: internal/common/grpc/grpcpool/pool.go
Comment (ClifHouck) [reply internal/common/grpc/grpcpool/pool.go]: Addressed in ad97bac

**Review by richscott — APPROVED** [review]

**Pull request merged**
